types/wlr_scene.c

```diff
@@ -133,6 +133,25 @@ void wlr_scene_node_place_below(struct wlr_scene_node *node,
 	wl_list_insert(sibling->state.link.prev, &node->state.link);
 }
 
+void wlr_scene_node_reparent(struct wlr_scene_node *node,
+		struct wlr_scene_node *new_parent) {
+	if (node->parent == new_parent) {
+		return;
+	}
+	/* Ensure that a node cannot become its own ancestor */
+	for (struct wlr_scene_node *ancestor = new_parent; ancestor != NULL;
+			ancestor = ancestor->parent) {
+		assert(ancestor != node);
+	}
+
+	wl_list_remove(&node->state.link);
+
+	node->parent = new_parent;
+	if (new_parent != NULL) {
```

Comment:
Do we want to allow a node without any parent? I think I'd rather not.

Reply:
OK. I wasn't sure if there might be a case where it would be convenient for a compositor to be able to detach a subtree for a period of time. I guess the existing code already requires you to attach non-root nodes to a tree when you create them, so it would be consistent with current invariants to prevent detaching.

(If the distinction between ROOT and TREE ever disappears, this might be worth revisiting.)

Reply:
Some code currently assumes each node has a root, especially with the damage tracking PRs. I'm not necessarily against allowing detached nodes, but need to be careful about edge cases, and I think `wlr_scene_node_set_enabled` can be used instead in most (?) cases.